### PR TITLE
Add serialization functionality for breakpoints

### LINKED ISSUE #707: Serialize Breakpoints to Projects
**Environment information**
* Operating System: Windows 10 Pro (20H2) 19042.804
* Cutter version: 1.12.0
* Obtained from:
  - [] Built from source
  - [x ] Downloaded release from Cutter website or GitHub 
  - [ ] Distribution repository
* File format: .exe

**Describe the bug**
Breakpoints are not saved.

I placed some breakpoints and saved my project, I loaded my project again, but I didn't find my saved breakpoints.

**To Reproduce**

1. Set a break point.
2. Save your project.
3. Close the program.
4. Load your saved project again.
5. Open breakpoint tab.

**Expected behavior**

Breakpoints must be loaded when I load my project again.

**Screenshots**

**Additional context**

I don't know If this is a bug or this feature not available yet.

### LINKED ISSUE #2735: Breakpoints are not saved
**Environment information**
* Operating System: windows 10
* Cutter version: 2.0.2
* File format: pe
 * Arch: x86
 * Type: EXEC (Executable file)

**Describe the bug**
Breakpoints are not saved to the project.

**To Reproduce**
1. Add a breakpoint.
2. Save the project.
3. Close the project.
4. Re open the same project.

**Expected behavior**
To have all breakpoint in the breakpoint window.

**What acutely happens**
No breakpoints in the breakpoint window.

### PR DESCRIPTION
**Pull request by DMaroo**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* Breakpoints now serialized and saved in project files
* You get them back where they were when you reopen a project

**Test plan**

Add tests once the serialization is approved and reviewed.

**Closing issues**

Closes #707.
Closes rizinorg/cutter#2735.

### CODE REVIEW

File: librz/bp/serialize_bp.c
Comment (ret2libc) [inline librz/bp/serialize_bp.c:23]: @thestr4ng3r is it preferred to save as json or as sdb directly?

Comment (DMaroo) [reply librz/bp/serialize_bp.c:23]: My motivation for doing this was basically the code from `serialize_flag.c`. Much of the code in this file is similar to that of serializing flag zones.

Also, my 2 cents: Serializing directly as sdb will probably write every attribute of every breakpoint as a new key value pair. When we first serialize to JSON and then store as sdb, the data looks prettier and less cluttered.

File: librz/bp/serialize_bp.c
Comment (ret2libc) [inline librz/bp/serialize_bp.c]: I saw other code using `RZ_KEY_PARSER_JSON`, is that possible here as well?

Comment (DMaroo) [reply librz/bp/serialize_bp.c]: I'll have a look. Thanks.

File: librz/bp/serialize_bp.c
Comment (thestr4ng3r) [reply librz/bp/serialize_bp.c:23]: As a rule of thumb, an sdb entry should correspond to a single "object", whatever that means in the specific case. Attributes should be contained in the value as json.
So here, creating one sdb entry per breakpoint with all info inside of it as json is the best approach in my opinion.
You just have to make sure `bp_item->name` really is a unique identifier for the bp, I don't know the breakpoint code well enough to tell whether it is without checking.

File: librz/bp/serialize_bp.c
Comment (DMaroo) [reply librz/bp/serialize_bp.c]: https://github.com/rizinorg/rizin/blob/dev/librz/analysis/serialize_analysis.c#L637-L664

Looks like we can also make a key parser to do this job, but I don't exactly know how helpful it will be. @thestr4ng3r Your thoughts, please.

File: librz/bp/serialize_bp.c
Comment (DMaroo) [reply librz/bp/serialize_bp.c:23]: Yes. Names aren't a unique identifier for breakpoints. I'll use addresses instead.

File: librz/bp/serialize_bp.c
Comment (thestr4ng3r) [reply librz/bp/serialize_bp.c]: Yes, the key parser should be used here. It's both more concise and faster since it avoids the `strcmp` chaining.

File: test/unit/test_serialize_bp.c
Comment (DMaroo) [inline test/unit/test_serialize_bp.c:16]: Tests are failing here. I think I am creating a breakpoint at a memory location non-existent in the current debug instance, however, I am not sure how to resolve this. Will require some help, @thestr4ng3r.

File: librz/bp/serialize_bp.c
Comment (ret2libc) [inline librz/bp/serialize_bp.c]:
```suggestion
		sdb_set(db, rz_strf(key, "0x%" PFMT64x, bp_item->addr), pj_string(j), 0);
```

**Comment by DMaroo** [comment]
The breakpoint loading doesn't seem to work. Everything looks correct and as expected in the `.rzdb` project file, but loading fails when trying to load the breakpoints. I suspect the error to be somewhere in the following lines: 

https://github.com/rizinorg/rizin/pull/1574/files#diff-e10dc9ab50cb92db60c5eaf3dd32a456141636e478a2e6b320f25d3c28892e08R82-R84

**Comment by DMaroo** [comment]
Breakpoint loading doesn't seem to work. I seem to be making some mistake here. From my analysis, the `bp_load_cb` callback function never gets called, and therefore `false` is returned by `rz_serialize_bp_load`. However, I am clearly able to see the saved breakpoints in the `.rzdb` project file. I suppose I must be missing to add or set something, but I am unable to figure it out.
cc @thestr4ng3r.

File: librz/bp/serialize_bp.c
Comment (XVilka) [inline librz/bp/serialize_bp.c:13]: Please add Doxygen

File: librz/debug/serialize_debug.c
Comment (XVilka) [inline librz/debug/serialize_debug.c:14]: Doxygen please

File: librz/debug/serialize_debug.c
Comment (XVilka) [inline librz/debug/serialize_debug.c:27]: Also Doxygen

File: librz/bp/serialize_bp.c
Comment (XVilka) [inline librz/bp/serialize_bp.c:302]: Doxygen

File: librz/bp/serialize_bp.c
Comment (XVilka) [inline librz/bp/serialize_bp.c:90]: Doxygen

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/project_migrate.c
Comment (DMaroo) [inline librz/core/project_migrate.c:244]: Will need to fill this commit hash once this PR gets merged into dev.

File: librz/core/project_migrate.c
Comment (XVilka) [inline librz/core/project_migrate.c:245]: Please also mention *exact* namespace where it goes.

**Review by XVilka — APPROVED** [review]
Overall looks good. 👍

**Review by ret2libc — CHANGES_REQUESTED** [review]
Just to block merging. I'd like to test it a bit before approving. Also, if @thestr4ng3r can give a look it would be great since he's the project expert :D

**Review by thestr4ng3r — APPROVED** [review]

**Comment by DMaroo** [comment]
> Just to block merging. I'd like to test it a bit before approving. Also, if @thestr4ng3r can give a look it would be great since he's the project expert :D

Have you tested it yet? Also, I could help in the testing, so please let me know in case there is need for some more testing.

**Comment by DMaroo** [comment]
Can this be merged? Thank you.

File: librz/bp/serialize_bp.c
Comment (ret2libc) [inline librz/bp/serialize_bp.c:256]: I was testing this on Windows with:
```
λ rizin-install\bin\rizin.exe -p rizin.rzdb -d rizin-install\bin\rizin.exe
DebugActiveProcess: (0x5) Access is denied.
DebugActiveProcess: (0x5) Access is denied.
WARNING: Breakpoint won't be placed since it's not in a valid map.
You can bypass this check by setting dbg.bpinmaps to false.
Spawned new process with pid 22532, tid = 21028
```
The breakpoints seem to work then, but I do wonder whether the restoring of the debug info happens at the right time... @thestr4ng3r idea? Is it alright as a first step?

Also, what about PIE? This is not going to work, is it?

**Comment by ret2libc** [comment]
> Can this be merged? Thank you.

Sorry for the delay, we were a bit busy finishing some bugs for the upcoming release. I hope to have this merged very soon!! Thanks for your patience

File: librz/bp/serialize_bp.c
Comment (DMaroo) [reply librz/bp/serialize_bp.c:256]: Thanks for testing on Windows. I'll check whether this works for PIE or not.

Comment (DMaroo) [reply librz/bp/serialize_bp.c:256]: The serialization works because (it seems) that base addresses are (maybe) stored in the project files. This seems to work for both PIE and no-PIE. I'll make sure that this is what is indeed happening (by inspecting the code), and the breakpoints serialization working is not just a coincidence.

Comment (ret2libc) [reply librz/bp/serialize_bp.c:256]: It is probably alright that it doesn't work, because IO/Bin is not properly serialized yet... let's see.

Comment (thestr4ng3r) [reply librz/bp/serialize_bp.c:256]: Right now, before IO/Bin serialization is implemented, the only available information is a filename. No base address or anything similar. But even with IO and Bin, that still won't give away much info about mapping in a debug process. That is a whole different story.

The idea right now would be to never load or save a project in debug mode, but only in static mode and then switch to debug with `ood`. Then the regular rebasing that happens should take care of relocating the breakpoints to wherever the file is statically loaded by default.

**Review by ret2libc — APPROVED** [review]

**Review by thestr4ng3r — APPROVED** [review]

**Pull request merged**
